### PR TITLE
[WIP; Early Preview; Discussion] New electron-based UI?

### PR DESCRIPTION
**Pull request by matthid**
I tried various times to contribute here but the amount of dependencies (especially the telerik stuff) and steps required to get it to run make it hard to do so.

I decided to start on my [own fork](https://github.com/matthid/CSGO-Demos-Manager) with merging existing contributions/open PRs and:
- [x] (Proof of concept) Replace the UI with a electron based app, currently written in F# (Currently very very limited and ugly)
- [x] (Proof of concept) Add an ASP.NET Core based backend again in F# for easy interop between backend/frontend
- [ ] Improve performance when downloading/analysing, and automating the process
- [x] add a build system (run a single command and implement in watch mode)
- [ ] Make the auto-updating installer for the electron app.

-> Basically the only thing that works in the new UI right now is the demo-list and downloading demos (but newly downloaded demos will not be shown in the list).

As I'm personally not a huge fan of throwing stuff away and rewriting everything, I decided to gradually improve the exiting code-base or change some internals where required in a non-breaking way. Currently I changed these:
- [x] `boiler.exe` is searched alongside the assembly in addition to the current working directory
- [x] Remove expression/reflection-based `Set` calls as those are a lot slower performance wise (and I saw these appear in the profiler) and switch to `nameof` and `[CallerMemberName]`
- [x] Fix a bug where `response.StatusCode` cannot be accessed after close in netcore applications (as an `ObjectDisposedException` is thrown)
- [x] Add `Microsoft.Extensions.Logging.Abstractions` to allow customized logging when outside the current WPF app.

Now the questions:
- Is it OK to work together on the core stuff like suggested in this PR (currently I added this repository as subtree and manage my own project files to compile for `netstandard`)? This is imho low-risk as improvements flow into both frontends and I can still use the old UI in case the new one never actually finishes ;).
- Is anyone interested in bringing the new electron UI on feature parity? @akiver Are you interested in updating the UI long-term in this repository (obviously a lot of work has to be done for this to be considered, just getting an early idea of thoughts here)

### CODE REVIEW

**Comment by JohBa** [comment]
Just to be clear, I (as a user and maybe a future contributor ;)) don't have a general problem with F#/Fable but I think F# might not boost the amount of contributions, similar to telerik.
I like the idea of a WebApp based demo manager.
I may be wrong about the contribution boost and would most likely also contribute if it runs with F# (mostly in frontend...)

**Comment by 0BuRner** [comment]
Hello,
Seems a good idea to me to make a new UI for this tool in particular regarding the usage of Telerik in the current version.
I also agree with @JohBa about using F# which might be a bad idea regarding contributions. Why not keep C# as main language ? What are the pros for using F# ?

**Comment by matthid** [comment]
Thank you both for the impressions!

> Why not keep C# as main language ? 

To be completely honest with both of you: I personally don't really care too much either way.
Doing lots of F# personally and knowing the [SAFE-Stack](https://safe-stack.github.io/) it was just the easiest to get started.
Additionally, I decided that it should be a web-based frontend with electron so C# is probably not an option for it in the frontend. Now the question is what alternatives you have as you already have to introduce a new stack of technology anyway. F# (or rather SAFE) just allows having a uniform experience at this point.

So, I would say it like this: If there are really some potential contributors like @0BuRner seems to be - We can use something else as a result of this discussion. But in my experience people just say they will contribute, but don't really do it in practice, so if the main part of the work is on me anyway, it should be clear that I choose the stuff I'm most familiar with. Just like @akiver did with the Telerik components (which again probably wasn't a bad decision from that standpoint).

To put it in other words: I'd definitely accept a PR changing the backend to C# for example (given the current POC state there probably isn't a lot of work involved). As stated above "the core" will probably stay in C# anyway. 

For the frontend we can discuss the alternatives as well:
- plain JS
- TypeScript
- something else or not electron based?

Again PRs changing stuff there can be discussed beforehand and probably are acceptable as long as they work. Just don't expect me to fiddle around with alternatives just because someone on the internet said he would then contribute, that did never work in my experience ;). But DO note that I really don't care too much about the technology. If it's something I haven't seen, I see it as a learning opportunity, so I'm willing to still push this forward (as long as it is not commercial).

The only other reason for me to change stuff would be if @akiver would consider accepting the UI changes (in addition to this PR) based on the technology decision because not splitting the ecosystem is always worth a bit of extra effort. But even then we first would need to get this PR in (which doesn't even involve any F# or new UI stuff).

Does that sound reasonable?

**Comment by JohBa** [comment]
I totally agree with you! As you know I am familiar with F# and Fable, so it won't be a blocker to me personally, and as you already noted just saying I want to contribute is not contributing yet.
My preference probably would be C# and TypeScript, but this is not my Repository nor my PR so this is just my two cents... ;)

**Comment by akiver** [comment]
Hi! Thank you for opening this PR which leads to discussing the future of the application 😃 

Before continuing I would like to say that the idea of using WEB technologies in the app is something that I thought a long time ago.
Actually, the version 2.0 embedded a web browser using [CefSharp](https://github.com/cefsharp/CefSharp) **only** for the heatmap feature that was using [heatmap.js](https://github.com/pa7/heatmap.js)
I removed this by using a C# heatmap implementation that solved my needs, but later I wanted to add features such as a 2D overview or stats graphs.
Sadly resources to do it quickly and easily in the WPF world are often commercial (like Telerik).
But yeah I like the idea :)

> I tried various times to contribute here but the amount of dependencies (especially the telerik stuff) and steps required to get it to run make it hard to do so.

Yes, the app has several packages dependencies but I think the only one that could be frustrating is Telerik because you need to have the binaries.
Otherwise, all dependencies are managed through NuGet which is what you may expect from a .NET application.
Your proposal includes 1 new language, several ecosystems like ASP, SAFE, webpack, Elmish... and a bridge to communicate between the "backend" and the "client".
This is impressive but it adds a lot of complexity, requires more knowledge to be able to understand how things work, reduces potential contributors and makes the process to upgrade the whole stack harder.  

The other thing which makes me think that your proposal is not the direction that I would like to see about the stack / code architecture (not the UI) for the next version is that I have a V3 on the road based on Electron.  
Around 2 years ago I started to think about the next major version and the 2 main things that I would like are:  
- Cross-platform app
- More contributors, which means a development platform available on all OS, no proprietary deps, a reasonable setup process and a language used by most devs. .NET is getting more and more open source but it still lacks the ability to create cross-platform **desktop** apps.

I looked into possible solutions that could do the job like QT (for the story the V1 was a QT application), Haxe, Electron... and I decided to go for Electron because I'm quite familiar with it and mainly because since it's basically a web browser in a window, all Node / web resources are available.  

The V3 is still under development, has missing/partial features compared to the V2 but I hope to release an "alpha" and make it public on GitHub in the next months.

Here is a quick preview of the app at its current state:

csgodm

It's written in TS, uses react / styled-components for the UI and requires a MySQL database.

To sum up I don't think this proposal is going to be the official future of the application.
But of course you can keep working on it and when you think it's ready, notify me and I will do everything to advertise your work and make it available from the website 😃  (which will be open source before the V3 BTW).

**Comment by matthid** [comment]
@akiver Sounds exactly like what I had in mind but with even more changes. So you basically rewrite everything and started from scratch (impressive!).

Any reason you didn't do it in the open so others could potentially join and help (like myself)?

When you release this and I'm continuing my project (which I doubt) I might change the frontend to TypeScript as well and share components so we can work together (Depending on how easy they are to integrate) Personally, I don't feel there are enough users or contributions to justify maintaining too much duplication.

Thanks for the detailed response!

**Comment by akiver** [comment]
> Any reason you didn't do it in the open so others could potentially join and help (like myself)?

I wanted to tackle some technical details related to Electron and create a base code before making it open source by myself.

I don't know how easy it will be to share components but that's a possibility.

**Comment by CorySanin** [comment]
When I first started looking through this thread, I thought to myself that it may just be easier and smarter to just redo the entire app in node and electron. The cross-platform aspect is a big deal, since CS:GO itself is also cross-platform. The only reason I didn't speak up was I thought the suggestion would've been a bit too ambitious. Glad to see you've already started doing exactly that! I look forward to contributing when the new version drops.
